Xamarin.Forms.Core/Element.cs

```diff
@@ -208,9 +208,37 @@ public Element Parent
 				OnParentSet();
 
 				OnPropertyChanged();
+
+				if (this.IsTemplateRoot)
+					this.TemplatedParent = value;
 			}
 		}
 
+		internal event EventHandler TemplatedParentChanged;
+		
+		BindableObject _templatedParent;
+		public BindableObject TemplatedParent
```

Comment:
No because `TemplatedParent` is populated for every `Element` created by a `ControlTemplate`. It's what `Binding.Apply` uses to locate the binding source in `TemplateBinding`s. So every `Element` potentially needs this member in order to eliminate the previous `async void` mechanism of crawling up the tree looking for the `IControlTemplated` every time a binding is applied.

And frankly if we were being thorough this would be populated for `Element`s made by `DataTemplate`s as well but I didn't want to go outside the scope of this issue.

Reply:
Oh, I see, that's great! The existing mechanism has been very "itchy" to me since I first saw it. Thanks for reply.